dm/simulator/internal/mcp/mcp.go

```diff
@@ -0,0 +1,132 @@
+// Copyright 2022 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package mcp defines the Modification Candidate Pool (MCP).
+package mcp
+
+import (
+	"math/rand"
+	"sync"
+	"time"
+
+	"github.com/pingcap/errors"
+	"github.com/pingcap/tiflow/dm/pkg/log"
+	"go.uber.org/zap"
+)
+
+var (
+	initRandSeedOnce sync.Once
+	gMCPRand         *rand.Rand
+	gMCPRandLock     sync.Mutex
+)
+
+func init() {
+	initRandSeedOnce.Do(func() {
+		gMCPRand = rand.New(rand.NewSource(time.Now().Unix()))
+	})
+}
+
+// ModificationCandidatePool is the core container storing all the current unique keys for a table.
+type ModificationCandidatePool struct {
+	sync.RWMutex
+	keyPool []*UniqueKey
+}
+
+// NewModificationCandidatePool create a new MCP.
+func NewModificationCandidatePool() *ModificationCandidatePool {
+	theKeyPool := make([]*UniqueKey, 0, 8192)
+	return &ModificationCandidatePool{
+		keyPool: theKeyPool,
+	}
+}
+
+// NextUK randomly picks a unique key in the MCP.
+func (mcp *ModificationCandidatePool) NextUK() *UniqueKey {
+	mcp.RLock()
+	defer mcp.RUnlock()
+	if len(mcp.keyPool) == 0 {
+		return nil
+	}
+	gMCPRandLock.Lock()
+	idx := gMCPRand.Intn(len(mcp.keyPool))
+	gMCPRandLock.Unlock()
+	return mcp.keyPool[idx] // pass by reference
+}
+
+// Len gets the current length of the MCP.
+func (mcp *ModificationCandidatePool) Len() int {
+	mcp.RLock()
+	defer mcp.RUnlock()
+	return len(mcp.keyPool)
+}
+
+// AddUK adds the unique key into the MCP.
+// It has side effect: the input UK's row ID will be changed.
+func (mcp *ModificationCandidatePool) AddUK(uk *UniqueKey) error {
+	mcp.Lock()
+	defer mcp.Unlock()
+	if len(mcp.keyPool)+1 > cap(mcp.keyPool) {
+		return errors.Trace(ErrMCPCapacityFull)
+	}
+	currentLen := len(mcp.keyPool)
+	uk.Lock()
+	uk.RowID = currentLen
+	uk.Unlock()
+	mcp.keyPool = append(mcp.keyPool, uk)
+	return nil
+}
+
+// DeleteUK deletes the unique key from the MCP.
+// It will get the row ID of the UK and delete the UK on that position.
+// If the actual value is different from the input UK, the element will still be deleted.
```

Comment:
So we can only pass in an index, not an *UniqueKey

Reply:
Originally, I planned to add the logic that if the RowID is invalid, the UK values are searched in MCP to get the actual index and delete.  So I kept the interface like this, in case in the further this logic is added back.